app/controllers/api_controller/authentication.rb

```diff
@@ -161,5 +149,50 @@ def api_action_details(collection, method, action)
     def api_token_mgr
       user_token_service.token_mgr
     end
+
+    def authenticate_with_user_token(x_auth_token)
+      @auth_token = x_auth_token
+      if !api_token_mgr.token_valid?(@module, @auth_token)
+        raise AuthenticationError, "Invalid Authentication Token #{@auth_token} specified"
+      else
+        @auth_user     = api_token_mgr.token_get_info(@module, @auth_token, :userid)
+        @auth_user_obj = userid_to_userobj(@auth_user)
+
+        unless request.headers['X-Auth-Skip-Token-Renewal'] == 'true'
+          api_token_mgr.reset_token(@module, @auth_token)
+        end
+
+        authorize_user_group(@auth_user_obj)
+        validate_user_identity(@auth_user_obj)
+        User.current_user = @auth_user_obj
+      end
+    end
+
+    def authenticate_with_system_token(x_miq_token)
+      @miq_token_hash = YAML.load(MiqPassword.decrypt(x_miq_token))
+
+      validate_system_token_server(@miq_token_hash[:server_guid])
+      validate_system_token_timestamp(@miq_token_hash[:timestamp])
+
+      @auth_user     = @miq_token_hash[:userid]
+      @auth_user_obj = userid_to_userobj(@auth_user)
+
+      authorize_user_group(@auth_user_obj)
+      validate_user_identity(@auth_user_obj)
+      User.current_user = @auth_user_obj
+    rescue => err
+      api_log_error("Authentication Failed with System Token\nX-MIQ-Token: #{x_miq_token}\nError: #{err}")
+      raise AuthenticationError, "Invalid System Authentication Token specified"
+    end
+
+    def validate_system_token_server(server_guid)
+      raise "Missing server_guid" if server_guid.blank?
+      raise "Invalid server_guid #{server_guid} specified" if MiqServer.where(:guid => server_guid).blank?
+    end
+
+    def validate_system_token_timestamp(timestamp)
+      raise "Missing timestamp" if timestamp.blank?
+      raise "Invalid timestamp #{timestamp} specified" if (Time.now.utc - timestamp).abs > 300
```

Comment:
There's gotta be a method that does this for you and doesn't make it look gross :)  Also, prefer `5.minutes` over `300`